guardian/manage-frontend
# Change Fastly to by default not cache unless route is explicitly added to allow-list

## What does this change?

_**Note this PR is just for audit purposes as we currently do not version control VCL, that is, the change was done directly via Fastly UI.**_

There exists a risk of serving personal data to wrong user due to having Fastly caching in-front of manage-frontend. This PR strengthens the safeguard by changing from deny-list principle to allow-list principle in Fastly, that is, for a route to become cacheable from now own it will have to be explicitly allowed. Previously, for example, adding a new route under say `/foo/bar` and accidentally placing this route above middleware that disables caching in `server.ts` might have resulted in caching of personal data.

VCL Before

```
 if ( req.url ~ "^/api/" ) {
      # Force PASS on all /api/* endpoints
      set beresp.ttl = 0s;
      set beresp.grace = 0s;
      return(pass);
  }
``` 

VCL After

```
 if ( req.url !~ "^/(static)" ) {
      # do not cache
      set beresp.ttl = 0s;
      set beresp.grace = 0s;
      return(pass);
  }
```

![image](https://user-images.githubusercontent.com/13835317/104719148-f15fee80-5723-11eb-8337-f5125560a01f.png)


From now one, to make the route `newroute` cachable 

1. Modify VCL condition to `req.url !~ "^/(static|newroute)"`
2. Add the route above `disableCache` middleware in `server.ts` or override `cache-control` headers 

[Fastly-&-Caching](https://github.com/guardian/manage-frontend/wiki/Fastly-&-Caching) corresponding wiki has been updated. (TODO)


## How to test

The change was tested in CODE by observing `x-cache` headers

![image](https://user-images.githubusercontent.com/13835317/104715422-7516dc80-571e-11eb-9266-51bf2bc6eb6e.png)

and Fastly stats dashboards 

![image](https://user-images.githubusercontent.com/13835317/104715663-bc9d6880-571e-11eb-9696-1cbed93d8d9e.png)


to see if `/static` routes are still cached whilst routes with personal data such as `/api/invoices` are not (`x-cache: MISS`)


## How can we measure success?

`x-cache: MISS` on all endpoints returning personal data.


## Have we considered potential risks?

* Changing default from allow-all to deny-all should strengthen the safety net in Fastly.
* There is a question of the value of having Fastly in place at all for something like manage-frontend since the purpose of the app is to manage personal accounts.
